### PR TITLE
zsh: fix nixUnstable completions

### PR DESCRIPTION
**Pull request by nrdxp**
###### Motivation for this change

The unstable nix completions work very well. Unfortunately, zsh users have so far been left to fend for themselves when it comes to getting them working. This fixes the situation by ~~bumping `nixUnstable` to a commit with the working zsh completion script, and~~ cleaning up a conflict with `nixStable` and `nix-zsh-completions`.

###### Things done

Tested on my local NixOS machine and it works

### CODE REVIEW

File: nixos/modules/programs/zsh/zsh.nix [hunk 1]
```diff
@@ -279,7 +279,7 @@ in
     environment.etc.zinputrc.source = ./zinputrc;
 
     environment.systemPackages = [ pkgs.zsh ]
-      ++ optional cfg.enableCompletion pkgs.nix-zsh-completions;
+      ++ optional (cfg.enableCompletion && ! lib.versionAtLeast (lib.getVersion config.nix.package) "2.4pre") pkgs.nix-zsh-completions;
```
Comment (SuperSandro2000) [inline nixos/modules/programs/zsh/zsh.nix:282]:
```suggestion
      ++ optional (cfg.enableCompletion && !lib.versionAtLeast (lib.getVersion config.nix.package) "2.4pre") pkgs.nix-zsh-completions;
```

Comment (nrdxp) [reply nixos/modules/programs/zsh/zsh.nix]: I added a commit with a pass of nixpkgs-fmt on the module.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 128649` at c49209fe run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built successfully:</summary>
  <ul>
    <li>hydra-unstable</li>
    <li>nix-update</li>
    <li>nixUnstable</li>
    <li>nixpkgs-review</li>
  </ul>
</details>

File: nixos/modules/programs/zsh/zsh.nix [hunk 2]
```diff
@@ -279,14 +281,18 @@ in
     environment.etc.zinputrc.source = ./zinputrc;
 
     environment.systemPackages = [ pkgs.zsh ]
-      ++ optional cfg.enableCompletion pkgs.nix-zsh-completions;
+      ++ optional
+      (cfg.enableCompletion
+        && ! lib.versionAtLeast (lib.getVersion config.nix.package) "2.4pre")
```
Comment (SuperSandro2000) [inline nixos/modules/programs/zsh/zsh.nix:286]:
```suggestion
        && !lib.versionAtLeast (lib.getVersion config.nix.package) "2.4pre")
```

Comment (nrdxp) [reply nixos/modules/programs/zsh/zsh.nix]: added in a force push

**Comment by Ma27** [comment]
Would it make sense to still install `nix-zsh-completions` for Nix 2.4, but to remove the completions for the `nix`-cmd from the derivation? Because it seems still useful for e.g. `nix-build` etc.

**Comment by nrdxp** [comment]
is there a simple filter function for derivations, or would I need to do a simple `mkDerivation`?

**Comment by Ma27** [comment]
> nvm, I'll just overrideAttrs and add a postInstall phase to remove it

Yup, that's what I've done in an overlay locally as well :) 

Btw as you seem to be using Nix Unstable completions as well, you may want to try out https://github.com/NixOS/nix/pull/4980 (basically support for completing `nix build -f . foo<tab>`). It shouldn't be added to this PR yet and I won't give safety guarantees for now as the patch is only a few hours old, but I'd be happy to get some feedback on this :)

**Comment by nrdxp** [comment]
> Btw as you seem to be using Nix Unstable completions as well, you may want to try out [NixOS/nix#4980](https://github.com/NixOS/nix/pull/4980) (basically support for completing `nix build -f . foo<tab>`).

I'll be glad to make a patch and see if I can get it working on my local machine. I am also currently using [nix-dram](https://github.com/nrdxp/nix-dram) so hopefully they don't conflict. Thanks for the tip!

Also, I just submitted a commit to implement your suggestion :+1:

**Comment by nrdxp** [comment]
Now that #129574 has been merged, I went ahead a removed the commit to bump the `nixUnstable` version, and this should be a lot less scary to merge :smile:

**Comment by Ma27** [comment]
@nrdxp do you think it'd make sense to backport this? :)

**Comment by nrdxp** [comment]
@Ma27, I think so, assuming the above mentioned nixUnstable bump on which this depends is also backported.

**Comment by Ma27** [comment]
It is :)

**Review by jonringer — APPROVED** [review]

**Pull request merged**